system/util-linux/text-utils/bits.1
'\" t
.\"     Title: bits
.\"    Author: [see the "AUTHOR(S)" section]
.\" Generator: Asciidoctor 2.0.26
.\"      Date: 2026-06-16
.\"    Manual: User Commands
.\"    Source: util-linux 2.42.2
.\"  Language: English
.\"
.TH "BITS" "1" "2026-06-16" "util\-linux 2.42.2" "User Commands"
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.ss \n[.ss] 0
.nh
.ad l
.de URL
\fI\\$2\fP <\\$1>\\$3
..
.als MTO URL
.if \n[.g] \{\
.  mso www.tmac
.  am URL
.    ad l
.  .
.  am MTO
.    ad l
.  .
.  LINKSTYLE blue R < >
.\}
.SH "NAME"
bits \- convert bit masks or lists from/to various formats
.SH "SYNOPSIS"
.sp
\fBbits\fP [\fB\-h\fP] [\fB\-V\fP] [\fB\-w\fP \fInumber\fP] [\fImode\fP] [\fImask\fP|\fIlist\fP].\|.\|.
.SH "DESCRIPTION"
.sp
The \fBbits\fP utility converts between bit masks and bit lists.
It supports combining multiple masks or lists using bitwise operations.
.SH "POSITIONAL ARGUMENTS"
.sp
\fImask\fP
.RS 4
A set of bits specified as a hexadecimal mask value
(for example: \fB0xeec2\fP).
.RE
.sp
\fIlist\fP
.RS 4
A set of bits specified as a comma\-separated list of bit IDs
(for example: \fB1,5,29,32\fP).
.RE
.sp
If no argument is specified, the sets of bits will be read from
standard input, one group per line.
.sp
Consecutive IDs can be compressed as ranges
(for example: \fB5,6,7,8,9,10\fP \(-> \fB5\-10\fP).
.sp
Optionally, if an argument starts with a comma, it will be
parsed as a single hexadecimal mask split in 32\-bit groups
(for example: \fB,00014000,00000000,00020000\fP \(-> \fB17,78,80\fP).
.sp
By default, all groups will be OR\(cqed together. If a group has one of
the prefixes \fB&\fP, \fB^\fP, or \fB~\fP, it will be combined with the
resulting mask using a different binary operation:
.sp
\fB&\fP\fImask\fP|\fB&\fP\fIlist\fP
.RS 4
The group will be combined with a binary AND operation.
That is: all bits that are set to 1 in the group AND in the combined groups
so far will be preserved as 1. All other bits will be reset to 0.
.RE
.sp
\fB^\fP\fImask\fP|\fB^\fP\fIlist\fP
.RS 4
The group will be combined with a binary XOR operation.
That is: all bits that are set to 1 in the group AND to 0 in the combined
groups so far (or the other way around) will be set to 1.
Bits that are set both to 1 or both to 0 will be reset to 0.
.RE
.sp
\fB~\fP\fImask\fP|\fB~\fP\fIlist\fP
.RS 4
All bits set to 1 in the group will be cleared (reset to 0) in the combined
groups so far.
.RE
.SH "OPTIONS"
.sp
\fB\-w\fP \fInumber\fP, \fB\-\-width\fP \fInumber\fP
.RS 4
The maximum number of bits in the masks handled by \fBbits\fP.
The default is \fB8192\fP. Any bit larger than this number will be truncated.
.RE
.sp
\fB\-h\fP, \fB\-\-help\fP
.RS 4
Display help text and exit.
.RE
.sp
\fB\-V\fP, \fB\-\-version\fP
.RS 4
Display version and exit.
.RE
.SH "CONVERSION MODE"
.sp
One of the following conversion modes can be specified. If not specified, it
defaults to \fB\-m\fP, \fB\-\-mask\fP.
.sp
\fB\-b\fP, \fB\-\-binary\fP
.RS 4
Print the combined arguments as a binary mask value.
.RE
.sp
\fB\-g\fP, \fB\-\-grouped\-mask\fP
.RS 4
Print the combined arguments as a hexadecimal mask value
in 32\-bit comma\-separated groups.
.RE
.sp
\fB\-l\fP, \fB\-\-list\fP
.RS 4
Print the combined arguments as a list of bit IDs.
Consecutive IDs are compressed to ranges.
.RE
.sp
\fB\-m\fP, \fB\-\-mask\fP
.RS 4
Print the combined arguments as a hexadecimal mask value (default).
.RE
.SH "EXAMPLES"
.sp
.if n .RS 4
.nf
.fam C
~$ bits \-\-mask 4,5\-8 16,30
0x400101f0

~$ bits \-\-list 0xeec2
1,6,7,9\-11,13\-15

~$ bits \-\-binary 4,5\-8 16,30
0b100_0000_0000_0001_0000_0001_1111_0000

~$ bits \-\-list ,00300000,03000000,30000003
0,1,28,29,56,57,84,85

~$ bits \-\-list 1,2,3,4 ~3\-10
1,2

~$ bits \-\-list 1,2,3,4 ^3\-10
1,2,5\-10

~$ bits \-\-grouped\-mask 2,22,74,79
8400,00000000,00400004

~$ bits \-\-width 64 \-\-list 2,22,74,79
2,22
.fam
.fi
.if n .RE
.SH "AUTHORS"
.sp
Robin Jarry.
.SH "REPORTING BUGS"
.sp
For bug reports, use the \c
.URL "https://github.com/util\-linux/util\-linux/issues" "issue tracker" "."
.SH "AVAILABILITY"
.sp
The \fBbits\fP command is part of the util\-linux package which can be downloaded from \c
.URL "https://www.kernel.org/pub/linux/utils/util\-linux/" "Linux Kernel Archive" "."

system/util-linux/text-utils/bits.c
/*
 * SPDX-License-Identifier: GPL-2.0-or-later
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * Copyright (c) 2024 Robin Jarry
 *
 * bits - convert bit masks from/to various formats
 */

#include <errno.h>
#include <getopt.h>
#include <sched.h>
#include <stdbool.h>
#include <stdint.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>

#include "c.h"
#include "closestream.h"
#include "cpuset.h"
#include "nls.h"
#include "strutils.h"
#include "strv.h"
#include "xalloc.h"

static void parse_mask_or_list(const char *cmdline_arg,
		cpu_set_t *all_bits, size_t width)
{
	cpu_set_t *bits, *copy;
	char bitwise_op = '|';
	const char *arg;
	size_t size, n;

	arg = cmdline_arg;

	/* strip optional operator first */
	if (ul_startswith(arg, "&")) {
		bitwise_op = '&';
		arg++;
	} else if (ul_startswith(arg, "^")) {
		bitwise_op = '^';
		arg++;
	} else if (ul_startswith(arg, "~")) {
		bitwise_op = '~';
		arg++;
	} else if (ul_startswith(arg, "|")) {
		arg++;
	}

	bits = cpuset_alloc(width, &size, NULL);
	if (bits == NULL)
		errx(EXIT_FAILURE, _("error: cannot allocate bit mask"));

	if (ul_startswith(arg, ",") || ul_startswith(arg, "0x")) {
		if (ul_startswith(arg, ","))
			arg++;
		if (cpumask_parse(arg, bits, size) < 0)
			errx(EXIT_FAILURE, _("error: invalid bit mask: %s"), cmdline_arg);
	} else {
		if (cpulist_parse(arg, bits, size, 1) < 0)
			errx(EXIT_FAILURE, _("error: invalid bit list: %s"), cmdline_arg);
	}

	/* truncate all bits beyond the requested mask size */
	for (n = cpuset_nbits(size) - 1; n >= width; n--)
		CPU_CLR_S(n, size, bits);

	copy = cpuset_alloc(width, &size, NULL);
	if (copy == NULL)
		errx(EXIT_FAILURE, _("error: cannot allocate bit mask"));
	memcpy(copy, all_bits, size);

	switch (bitwise_op) {
	case '&':
		CPU_AND_S(size, all_bits, copy, bits);
		break;
	case '|':
		CPU_OR_S(size, all_bits, copy, bits);
		break;
	case '^':
		CPU_XOR_S(size, all_bits, copy, bits);
		break;
	case '~':
		for (n = 0; n < width; n++) {
			if (CPU_ISSET_S(n, size, bits))
				CPU_CLR_S(n, size, all_bits);
		}
		break;
	}

	cpuset_free(bits);
	cpuset_free(copy);
}

static size_t num_digits(size_t value)
{
	size_t digits = 1;
	while (value > 9) {
		digits++;
		value /= 10;
	}
	return digits;
}

enum output_mode {
	MODE_BINARY,
	MODE_GROUPED_MASK,
	MODE_LIST,
	MODE_MASK,
};

static void print_bits(cpu_set_t *bits, size_t width, enum output_mode mode)
{
	const size_t size = CPU_ALLOC_SIZE(width);
	bool started = false;
	char *buf = NULL;
	size_t buf_size;
	ssize_t n = 0;

	if (CPU_COUNT_S(size, bits) == 0) {
		switch (mode) {
		case MODE_MASK:
			printf("0x0\n");
			break;
		case MODE_GROUPED_MASK:
			printf("0\n");
			break;
		case MODE_BINARY:
			printf("0b0\n");
			break;
		case MODE_LIST:
			break;
		}
		return;
	}

	switch (mode) {
	case MODE_MASK:
		/* fit 4 bits per character plus terminating nul byte */
		buf_size = ((cpuset_nbits(size) + 3) / 4) + 1;
		buf = xmalloc(buf_size);
		cpumask_create(buf, buf_size, bits, size);

		/* strip leading zeroes */
		while (buf[n] == '0')
			n++;
		printf("0x%s\n", buf + n);
		break;

	case MODE_GROUPED_MASK:
		/* fit 4 bits per character plus terminating nul byte */
		buf_size = ((cpuset_nbits(size) + 3) / 4) + 1;
		buf = xmalloc(buf_size);
		cpumask_create(buf, buf_size, bits, size);

		/* strip leading zeroes */
		while (buf[n] == '0')
			n++;

		while (buf[n] != '\0') {
			if (started && (n % 8) == 0)
				printf(",");
			if (buf[n] != '0')
				started = true;
			printf("%c", buf[n]);
			n++;
		}
		printf("\n");
		break;

	case MODE_BINARY:
		printf("0b");
		for (n = width - 1; n >= 0; n--) {
			if (started && ((n + 1) % 4) == 0)
				printf("_");
			if (CPU_ISSET_S(n, size, bits)) {
				started = true;
				printf("1");
			} else if (started) {
				printf("0");
			}
		}
		printf("\n");
		break;

	case MODE_LIST:
		/* Maximum number of digits (larger bit number) plus 1
		 * to account for a separating comma, times the number of bits
		 * set to 1. */
		buf_size = (num_digits(width - 1) + 1) * CPU_COUNT_S(size, bits);
		buf = xmalloc(buf_size);
		cpulist_create(buf, buf_size, bits, size);
		printf("%s\n", buf);
		break;
	}

	free(buf);
}

static void __attribute__((__noreturn__)) usage(void)
{
	fputs(USAGE_HEADER, stdout);
	fprintf(stdout, _(" %s [options] [<mask_or_list>...]\n"),
		program_invocation_short_name);

	fputs(USAGE_SEPARATOR, stdout);
	fputsln(_("Convert bit masks from/to various formats."), stdout);

	fputs(USAGE_ARGUMENTS, stdout);
	fputsln(_(" <mask_or_list>      bits specified as a hex mask (e.g. 0xeec2)\n"
		  "                       or as a comma-separated list of bit IDs"),
		stdout);
	fputs(USAGE_SEPARATOR, stdout);
	fputsln(_(" If not specified, arguments will be read from stdin."),
		stdout);

	fputs(USAGE_OPTIONS, stdout);
	fprintf(stdout, USAGE_HELP_OPTIONS(21));
	fputsln(_(" -w <num>, --width <num>\n"
	          "                     maximum width of bit masks (default 8192)"),
		stdout);

	fputs(_("\nOutput modes:\n"), stdout);
	fputsln(_(" -m, --mask          display bits as a hex mask value (default)"),
		stdout);
	fputsln(_(" -g, --grouped-mask  display bits as a hex mask value in 32bit\n"
		  "                       comma separated groups"), stdout);
	fputsln(_(" -b, --binary        display bits as a binary mask value"),
		stdout);
	fputsln(_(" -l, --list          display bits as a compressed list of bit IDs"),
		stdout);

	fprintf(stdout, USAGE_MAN_TAIL("bits(1)"));
	exit(EXIT_SUCCESS);
}

int main(int argc, char **argv)
{
	enum output_mode mode = MODE_MASK;
	char **stdin_lines = NULL;
	cpu_set_t *bits = NULL;
	size_t width = 8192;
	size_t alloc_size;
	int c;

#define FLAGS "Vhw:mgbl"
	static const struct option longopts[] = {
		{ "version",      no_argument,       NULL, 'V' },
		{ "help",         no_argument,       NULL, 'h' },
		{ "width",        required_argument, NULL, 'w' },
		{ "mask",         no_argument,       NULL, 'm' },
		{ "grouped-mask", no_argument,       NULL, 'g' },
		{ "binary",       no_argument,       NULL, 'b' },
		{ "list",         no_argument,       NULL, 'l' },
		{ NULL,           0,                 NULL,  0  }
	};

	setlocale(LC_ALL, "");
	bindtextdomain(PACKAGE, LOCALEDIR);
	textdomain(PACKAGE);
	close_stdout_atexit();

	while ((c = getopt_long(argc, argv, FLAGS, longopts, NULL)) != -1) {
		switch (c) {
		case 'm':
			mode = MODE_MASK;
			break;
		case 'g':
			mode = MODE_GROUPED_MASK;
			break;
		case 'b':
			mode = MODE_BINARY;
			break;
		case 'l':
			mode = MODE_LIST;
			break;
		case 'w':
			/* allow up to 128k masks */
			width = str2unum_or_err(optarg,
				10, _("invalid --width"), 128 * 1024);
			if (width == 0)
				errx(EXIT_FAILURE, _("invalid --width"));
			break;
		case 'V':
			print_version(EXIT_SUCCESS);
		case 'h':
			usage();
		default:
			errtryhelp(EXIT_FAILURE);
		}
	}

	argc -= optind;
	argv += optind;
	if (argc == 0) {
		/* no arguments provided, read lines from stdin */
		char *buf = NULL;
		size_t bufsz = 0;

		while (getline(&buf, &bufsz, stdin) != -1) {
			/* strip LF, CR, CRLF, LFCR */
			rtrim_whitespace((unsigned char *)buf);
			if (ul_strv_push(&stdin_lines, xstrdup(buf)) < 0)
				errx(EXIT_FAILURE, _("cannot allocate memory"));
		}
		free(buf);

		argc = ul_strv_length(stdin_lines);
		argv = stdin_lines;
	}

	bits = cpuset_alloc(width, &alloc_size, NULL);
	if (bits == NULL)
		errx(EXIT_FAILURE, _("cannot allocate memory"));

	/* start off with all bits set to 0 */
	memset(bits, 0, alloc_size);

	for (; argc > 0; argc--, argv++)
		parse_mask_or_list(*argv, bits, width);

	ul_strv_free(stdin_lines);

	print_bits(bits, width, mode);

	cpuset_free(bits);

	return EXIT_SUCCESS;
}
